examples/core/rgcn/hetero_rgcn.py

```diff
@@ -0,0 +1,654 @@
+                for ntype in self.ntypes
+            }
+        )
+
+        self.dropout = nn.Dropout(dropout)
+        # Initialize parameters of the model.
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        for layer in self.weight.values():
+            layer.reset_parameters()
+
+        for layer in self.loop_weights.values():
+            layer.reset_parameters()
+
+    def forward(self, g, inputs):
+        """
+        Parameters
+        ----------
+        g : DGLGraph
+            Input graph.
+        inputs : dict[str, torch.Tensor]
+            Node feature for each node type.
+
+        Returns
+        -------
+        dict[str, torch.Tensor]
+            New node features for each node type.
+        """
+        # Create a deep copy of the graph g with features saved in local
+        # frames to prevent side effects from modifying the graph.
+        g = g.local_var()
+
+        # Create a dictionary of weights for each relationship. The weights
+        # are retrieved from the Linear layers defined earlier.
+        weight_dict = {
+            rel_name: {"weight": self.weight[rel_name].weight.T}
+            for rel_name in self.relation_names
+        }
+
+        # Create a dictionary of node features for the destination nodes in
+        # the graph. We slice the node features according to the number of
+        # destination nodes of each type. This is necessary because when
+        # incorporating the effect of self-loop edges, we perform computations
+        # only on the destination nodes' features. By doing so, we ensure the
+        # feature dimensions match and prevent any misuse of incorrect node
+        # features.
+        inputs_dst = {
+            k: v[: g.number_of_dst_nodes(k)] for k, v in inputs.items()
+        }
+
+        # Apply the convolution operation on the graph. mod_kwargs are
+        # additional arguments for each relation function defined in the
+        # HeteroGraphConv. In this case, it's the weights for each relation.
+        hs = self.conv(g, inputs, mod_kwargs=weight_dict)
+
+        def _apply(ntype, h):
+            h = h + self.loop_weights[ntype](inputs_dst[ntype])
+            if self.activation:
+                h = self.activation(h)
+            return self.dropout(h)
+
+        # Apply the function defined above for each node type. This will update
+        # the node features using the self-loop weights, apply the activation
+        # function and dropout.
+        return {ntype: _apply(ntype, h) for ntype, h in hs.items()}
+
+
+class EntityClassify(nn.Module):
+    def __init__(self, g, in_size, out_dim):
+        super(EntityClassify, self).__init__()
+        self.in_size = in_size
+        self.hidden_size = 64
+        self.out_dim = out_dim
+
+        # Generate and sort a list of unique edge types from the input graph.
+        # eg. ['writes', 'cites']
+        self.relation_names = list(set(g.etypes))
```

Comment:
Could you explain why we don't use canonical etypes here, e.g., ('author', 'write', 'paper')?

Reply:
Actually, I also don'k know why we train for unique `g.etypes` instead of `g.canonical_etypes`.